### PR TITLE
update events configured to watch specific fields

### LINKED ISSUE #203: Make update events configurable to watch only specific fields
**Is your feature request related to a problem? Please describe.**
Currently, BotKube watches and sends notifications only about updates in `spec` of a resource when configured to watch `update` events
It should be configurable which resource fields we want watch and get notifications for when updated

**Describe the solution you'd like**
We can change `update` fields in configuration to -
```
update:
  fields:
    - spec
    - metadata
    - status
  includeDiff: true
```

### PR DESCRIPTION
**Pull request by Surbhidongaonkar**
##### ISSUE TYPE
 - Feature Pull Request

##### SUMMARY
This commit  makes botkube notify about specific field updates. The updates can be in any of the spec, metadata or status. 
Botkube should notify about this according to users specifications.

Fixes #203

### CODE REVIEW

**Comment by Surbhidongaonkar** [comment]
Fixing the failed checks, Will Update once done.

**Comment by Surbhidongaonkar** [comment]
> Fixing the failed checks, Will Update once done.

Fixed by modifying cmp_test.go

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -163,7 +163,14 @@ func sendEvent(obj, oldObj interface{}, c *config.Config, notifiers []notify.Not
...
 	if eventType == config.UpdateEvent {
-		updateMsg := utils.Diff(oldObj, obj)
+		updatesettingAll, existsAll := utils.AllowedUpdateEventsMap[utils.KindNS{Resource: kind, Namespace: "all"}]
+		updatesettingNS, existsNS := utils.AllowedUpdateEventsMap[utils.KindNS{Resource: kind, Namespace: objectMeta.Namespace}]
+		var updateMsg string
+		if existsAll {
+			updateMsg = utils.Diff(oldObj, obj, updatesettingAll)
+		} else if existsNS {
+			updateMsg = utils.Diff(oldObj, obj, updatesettingNS)
+		}
```
Comment (PrasadG193) [inline pkg/controller/controller.go]: We can avoid code duplication here.
```
// Check if all namespaces allowed
updateSetting, exist := utils.AllowedUpdateEventsMap[utils.KindNS{Resource: kind, Namespace: "all"}]
fi !exist {
    // Check if specified namespace is allowed
    updateSetting, exist = utils.AllowedUpdateEventsMap[utils.KindNS{Resource: kind, Namespace: objectMeta.Namespace}]
}
if exist {
    updateMsg = utils.Diff(oldObj, obj, updatesettingNS
}
```

Comment (Surbhidongaonkar) [reply pkg/controller/controller.go]: Hi @PrasadG193, Made the required changes. Please review.

**Comment by PrasadG193** [comment]
@Surbhidongaonkar could you please rebase with develop?

File: pkg/utils/cmp.go [hunk 2]
```diff
@@ -5,44 +5,51 @@ import (
...
 }
 
 // Report custom implements Reporter interface
-func (r *SpecDiffReporter) Report(rs cmp.Result) {
+func (r *DiffReporter) Report(rs cmp.Result) {
 	if !rs.Equal() {
 		vx, vy := r.path.Last().Values()
 		path := fmt.Sprintf("%#v", r.path)
-		if ok := strings.Contains(path, ".Spec."); ok {
+		if ok := strings.Contains(path, "."+strings.Title(r.field)); ok {
```
Comment (PrasadG193) [inline pkg/utils/cmp.go]: I think we should support regex/jsonpath here.
Suppose I am interested only in change in container status of Pod, so I can just set `status.containerStatuses[*].ready` in updateSetting and that should work.

Comment (PrasadG193) [reply pkg/utils/cmp.go]: I have some working code changes for jsonpath on my local repo. I'll cherry-pick the commits and  raise a new PR

**Comment by PrasadG193** [comment]
@Surbhidongaonkar I have pushed few commit. Can you please refactor diff_test.go accordingly?

**Comment by PrasadG193** [comment]
Screenshots:

image

image

**Comment by Surbhidongaonkar** [comment]
> @Surbhidongaonkar I have pushed few commit. Can you please refactor diff_test.go accordingly?

sure! Will do.

**Comment by Surbhidongaonkar** [comment]
> @Surbhidongaonkar I have pushed few commit. Can you please refactor diff_test.go accordingly?

Updated diff_test.go.

**Pull request merged**